components/services/brave_wallet/public/cpp/third_party_service.cc

```diff
@@ -0,0 +1,142 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/services/brave_wallet/public/cpp/third_party_service.h"
+
+#include <utility>
+
+#include "base/functional/callback.h"
+#include "base/no_destructor.h"
+#include "brave/components/services/brave_wallet/public/cpp/third_party_service_launcher.h"
+#include "brave/components/services/brave_wallet/public/mojom/filecoin_utility.mojom.h"
+#include "build/build_config.h"
+
+namespace brave_wallet {
+
+namespace {
+
+constexpr base::TimeDelta kServiceProcessIdleTimeout{base::Seconds(5)};
+
+template <typename T, typename V>
+class Request : public base::RefCounted<Request<T, V>> {
+ public:
+  explicit Request(ThirdPartyService::ResultCallback<V> callback)
+      : callback_(std::move(callback)) {}
+
+  Request(const Request&) = delete;
+  Request& operator=(const Request&) = delete;
+
+  mojo::Remote<T>& remote() { return remote_; }
+  ThirdPartyService::ResultCallback<V>& callback() { return callback_; }
+
+  // Creates a pipe and binds it to the remote(), and sets up the
+  // disconnect handler to invoke callback() with an error.
+  mojo::PendingReceiver<T> BindRemote() {
+    auto receiver = remote_.BindNewPipeAndPassReceiver();
+    remote_.set_disconnect_handler(
+        base::BindOnce(&Request::OnRemoteDisconnected, this));
+    return receiver;
+  }
+
+  void OnResult(const absl::optional<V>& result) {
+    if (!callback()) {
+      return;
+    }
+
+    // Copy the callback onto the stack before resetting the Remote, as that may
+    // delete |this|.
+    auto local_callback = std::move(callback());
+
+    // Reset the |remote_| since we aren't using it again and we don't want it
+    // to trip the disconnect handler. May delete |this|.
+    remote_.reset();
+
+    // We run the callback after reset just in case it does anything funky like
+    // spin a nested RunLoop.
+    std::move(local_callback).Run(std::move(result));
+  }
+
+ private:
+  friend class base::RefCounted<Request>;
+
+  ~Request() = default;
+
+  void OnRemoteDisconnected() {
+    if (callback()) {
+      std::move(callback()).Run(absl::nullopt);
+    }
+  }
+
+  mojo::Remote<T> remote_;
+  ThirdPartyService::ResultCallback<V> callback_;
+};
+
+}  // namespace
+
+// static
+ThirdPartyService& ThirdPartyService::Get() {
+  static base::NoDestructor<ThirdPartyService> service;
```

Comment:
Make it this way cuz I think we could just have one instance (and one process) to be shared inside wallet component and not have multiple processes like data_decoder.